### ISSUE TITLE
twisted.words irc server does not work with gaim

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2385]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#2385|
|Type|defect|
|Created|2007-01-10 21:16:50Z|
|Branch|https://github.com/twisted/twisted/tree/pidgin-2385-3|

When I attempt to connect, regardless of whether I specify a password or not, gaim hangs forever and doesn't consider me logged in.

<details><summary>Searchable metadata</summary>

```
trac-id__2385 2385
type__defect defect
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__branches_pidgin_2385_3 branches/pidgin-2385-3
branch_author__tenth tenth
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1168463810000000 1168463810000000
changetime__1235351480000000 1235351480000000
version__None None
owner__tenth tenth

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

I checked with ethereal.  The login seems to be succeeding but the Twisted Words server doesn't send very many codes in its greeting.  I suppose gaim might be waiting for something which it doesn't send before considering login complete.

**Comment by twisted-trac** [comment]
|[<img alt="tenth10th's avatar" src="https://avatars.githubusercontent.com/u/854235?s=50" width="50" height="50">](https://github.com/tenth10th)<a name="note_2"></a>|@tenth10th commented|
|-|-|

Itamar and I discovered that including the MOTD lines:
        ("375", ":- %(serviceName)s Message of the Day - "),
        ("372", ":- The message of the day is: THERE IS NO MESSAGE OF THE DAY. - "),
        ("376", ":End of /MOTD command.")
as part of twisted.words.service.IRCUser._welcomeMessages, immediately after the RPL_MYINFO line, is consistent with FreeNode server behavior, and is more acceptable to GAIM.

**Comment by twisted-trac** [comment]
|[<img alt="tenth10th's avatar" src="https://avatars.githubusercontent.com/u/854235?s=50" width="50" height="50">](https://github.com/tenth10th)<a name="note_3"></a>|@tenth10th commented|
|-|-|

Actually, you don't even need the 372 (actual message of the day) line, just 375 (RPL_MOTDSTART) and then 376 (RPL_ENDOFMOTD). hooray.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Awesome, thanks for the investigation.  Now it should be easy for someone to adjust `IRCUser` to send these codes and make gaim/pidgin work.

**Comment by twisted-trac** [comment]
|[<img alt="tenth10th's avatar" src="https://avatars.githubusercontent.com/u/854235?s=50" width="50" height="50">](https://github.com/tenth10th)<a name="note_5"></a>|@tenth10th commented|
|-|-|

(In [25256]) Branching to 'pidgin-2385'

**Comment by twisted-trac** [comment]
|[<img alt="tenth10th's avatar" src="https://avatars.githubusercontent.com/u/854235?s=50" width="50" height="50">](https://github.com/tenth10th)|@tenth10th removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set owner to @tenth10th|
|-|-|

Here we go:

1.
```
-            "serviceName": self.hostname,
+            "serviceName": "test",
```

Why's this change?

2.
```
+        self.name = nickname;
```

Please remove the semi-colon.

3. One flake:
```
twisted/words/test/test_irc_service.py:8: 'time' imported but unused
```

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="tenth10th's avatar" src="https://avatars.githubusercontent.com/u/854235?s=50" width="50" height="50">](https://github.com/tenth10th)<a name="note_8"></a>|@tenth10th removed owner|
|-|-|

Itamar and I addressed Therve's review comments, and also added an actual test for Logins. As part of this, we also fixed some bugs found by the test (hostname is always "undefined", fake creation time, etc.)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Looking good.  A few simple things:

  1.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @tenth10th<br>@exarkun set status to `new`|
|-|-|

Let me try that again.

First, some simple things:

  1. In `service.py`, the new `serverInfo` attribute of the factory needs to be documented.  In particular, all the keys (and their associated values) which it will have need documentation.
  1. In `test_service.py`, the `_assertGreeting` method's new docstring has some trailing whitespace in it; it also documents `AssertionError` as an exception which may be raised, but I think it's actually `twisted.trial.unittest.FailTest`, or maybe `self.failureException`.
  1. Further down in that file, the new `messageType` parameter to `_response` is undocumented.
  1. `testNickServLogin`'s docstring also has trailing whitespace.  As long as you're modifying the function, you should rename it to comply with the naming convention - `test_nickServLogin`.
  1. In `test_irc_service.py`
    1. numerous lines end in trailing whitespace
    1. methods should be separated by two blank lines; many are, but setUp and the first test method following it are separated by only one blank line.
    1. The dates in the copyright statement should be made accurate - 2009, not 2001-2008.

Next.  The behavior actually being implemented is very bizarre.  Now the server sends the MOTD (except there's never an MOTD so it just sends two lines about how there's not an MOTD) whenever it receives `NICK`.  As a separate bug, the server allows multiple `NICK`s to be issued.  This might be a reasonable thing to allow, but it's not being handled properly now.  Please file a new ticket for dealing with that misbehavior.

So, anyway.  This isn't quite the behavior of other IRC servers (at least, not freenode).  Freenode won't send anything until it gets _both_ `USER` and `NICK`.  Then it sends the whole "greeting" sequence.  Then it uses NickServ to challenge you for a password.  It would probably be best to emulate this more closely.

Finally, I don't really understand a lot of the changes to `service.py`.  The `name` attribute, at least, seems necessary to make the test pass (although it wasn't before - I didn't bother to figure out why, but I'm annoyed that it's not obvious :/).  However, why are `nickname` and `realm` being handled differently now?  `realm` in particular is actually part of `IUser` and should probably remain that way unless there's a good reason to change it.

**Comment by twisted-trac** [comment]
|[<img alt="tenth10th's avatar" src="https://avatars.githubusercontent.com/u/854235?s=50" width="50" height="50">](https://github.com/tenth10th)<a name="note_11"></a>|@tenth10th removed owner|
|-|-|

Itamar and I have made the suggested docstring and whitespace fixes. The new ticket [#3664](https://github.com/twisted/twisted/issues/3664) addresses the MOTD ordering issue. As for nickname and realm, they were required previously, but not set until after login; This new order seems to make more sense. (For example, error messages early in the login process will now contain the correct hostname.)

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|z3p set owner to @tenth10th|
|-|-|

Conflict when merging w/ trunk.

**Comment by twisted-trac** [comment]
|[<img alt="tenth10th's avatar" src="https://avatars.githubusercontent.com/u/854235?s=50" width="50" height="50">](https://github.com/tenth10th)<a name="note_13"></a>|@tenth10th commented|
|-|-|

(In [26276]) Branching to 'pidgin-2385-2'

**Comment by twisted-trac** [comment]
|[<img alt="tenth10th's avatar" src="https://avatars.githubusercontent.com/u/854235?s=50" width="50" height="50">](https://github.com/tenth10th)<a name="note_14"></a>|@tenth10th commented|
|-|-|

(In [26277]) Branching to 'pidgin-2385-3'

**Comment by twisted-trac** [comment]
|[<img alt="tenth10th's avatar" src="https://avatars.githubusercontent.com/u/854235?s=50" width="50" height="50">](https://github.com/tenth10th)<a name="note_15"></a>|@tenth10th removed owner|
|-|-|

Merged forward (hopefully correctly this time!)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph set owner to @tenth10th|
|-|-|

Very minor changes:

  1. Since the structure of `serverInfo` is a bit silly (it's a dictionary, but it needs to have specific keys, it should really be a collection of attributes) it should be private pending some better structure that we want to publicly expose.  In other words, it should start with a single underscore, `_serverInfo`.
  2. A couple of blank lines have been incorrectly (I assume accidentally?) deleted which aren't really relevant to the implementation, and should be put back in.  Merged to trunk, these lines are:
    service.py:207
    service.py:350
    service.py:357
    service.py:888
  3. There is another problem, for which there should be a separate ticket.  AWAY isn't implemented on the server, so when I attempt to log in, I see this (thankfully, apparently harmless) traceback:
```
2009-02-22 19:26:06-0500 [IRCUser,0,127.0.0.1] Unhandled Error
	Traceback (most recent call last):
	  File "/home/glyph/Projects/Twisted/trunk/twisted/python/context.py", line 37, in callWithContext
	    return func(*args,**kw)
	  File "/home/glyph/Projects/Twisted/trunk/twisted/internet/selectreactor.py", line 146, in _doReadOrWrite
	    why = getattr(selectable, method)()
	  File "/home/glyph/Projects/Twisted/trunk/twisted/internet/tcp.py", line 459, in doRead
	    return self.protocol.dataReceived(data)
	  File "/home/glyph/Projects/Twisted/trunk/twisted/words/protocols/irc.py", line 172, in dataReceived
	    self.handleCommand(command, prefix, params)
	--- <exception caught here> ---
	  File "/home/glyph/Projects/Twisted/trunk/twisted/words/protocols/irc.py", line 184, in handleCommand
	    self.irc_unknown(prefix, command, params)
	  File "/home/glyph/Projects/Twisted/trunk/twisted/words/protocols/irc.py", line 191, in irc_unknown
	    raise NotImplementedError(command, prefix, params)
	exceptions.NotImplementedError: ('AWAY', '', [])
```
Other than that it looks good!  Please fix and merge.

**Comment by twisted-trac** [comment]
|[<img alt="tenth10th's avatar" src="https://avatars.githubusercontent.com/u/854235?s=50" width="50" height="50">](https://github.com/tenth10th)<a name="note_17"></a>|@tenth10th commented|
|-|-|

serverInfo has been marked private with an underscore, and the blank lines have been unremoved to clean up the diff in r26279. I have entered the AWAY implementation issue as ticket [#3668](https://github.com/twisted/twisted/issues/3668).

**Comment by twisted-trac** [comment]
|[<img alt="tenth10th's avatar" src="https://avatars.githubusercontent.com/u/854235?s=50" width="50" height="50">](https://github.com/tenth10th)<a name="note_18"></a>|@tenth10th set status to `closed`|
|-|-|

(In [26281]) Make twisted.words IRC server compatible with Pidgin.

author: itamar, tenth

reviewer: therve, exarkun, z3p, glyph

Fixes [#2385](https://github.com/twisted/twisted/issues/2385)

While there are other issues to be resolved with MOTD and AWAY behaviors 
as per [#3664](https://github.com/twisted/twisted/issues/3664) and [#3668](https://github.com/twisted/twisted/issues/3668) respectively, twisted.words now always sends an 
MOTD message, which is required for Pidgin to successfully connect to 
it. Also, the "nickname" and "realm" attributes are now initialized in 
time for them to appear in error messages during login.
